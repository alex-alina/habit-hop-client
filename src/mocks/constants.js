const storageJwtKey = 'jwt';
const mockJwt =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImRjM2MxMWJmLTU4NGMtNDA5MC1hYThiLTM5ZTQzY2ZhZmNhYiIsImlhdCI6MTY4MTQyMDcwMCwiZXhwIjoxNjgxNDI3OTAwfQ.sYclsZo_DO9zZ2375Kg-AiGZpiVvox1NojPK3eGj9k4';
const mockUserData = {
  createdAt: '2023-04-25T19:08:15.956Z',
  id: 'dc3c11bf-584c-4090-aa8b-39e43cfafcab',
  email: 'test@example.com',
  firstName: 'Billie',
  lastName: 'Test',
  lastLogin: '2023-04-25T19:08:15.956Z',
};

const mockGoalFormData = {
  goalDefinition:
    'Improve sleep quality by exercising and drinking less coffee',
  priority: 'main',
  startDate: '2023-03-23',
  endDate: '2023-06-16',
};

const mockGoalData1 = {
  endDate: '2023-05-06',
  goalDefinition:
    'Improve sleep quality by exercising and drinking less coffee',
  id: 'bec49b7a-7819-453b-9895-347e3982acc5',
  priority: 'main',
  startDate: '2023-04-08',
};

const mockGoalData2 = {
  endDate: '2023-05-06',
  goalDefinition:
    'I want to learn how to play guitar, and be able to play 3 songs by the end of the year',
  id: '44d96277-99ea-4a83-ab80-deed9148ddc1',
  priority: 'main',
  startDate: '2023-04-08',
};

const mockGoalData3 = {
  endDate: '2023-05-06',
  goalDefinition:
    'I want to learn how to bake muffins, cupcakes and birthday cakes by baking every weekend for six months',
  id: '1c8a4a14-7fed-4a91-b48b-a9290677043d',
  priority: 'main',
  startDate: '2023-04-08',
};

const editGoalData = {
  endDate: '2023-05-06',
  goalDefinition:
    'Train for a tripple marathon by running every day for 15km and going to the gym twice per week',
  id: 'bec49b7a-7819-453b-9895-347e3982acc5',
  priority: 'main',
  startDate: '2023-04-08',
};

const mockHabitFormData = {
  habitDescription: 'Exercise daily',
  habitType: 'develop',
  progressMetric: 'minutes',
};

const mockHabitData = {
  id: '66c8b782-6238-42a2-ab77-2c03d19b1d0f',
  habitDescription: 'Exercise daily',
  habitType: 'develop',
  progressMetric: 'minutes',
  goal: {
    id: 'bec49b7a-7819-453b-9895-347e3982acc5',
    goalDefinition:
      'Improve sleep quality by exercising and drinking less coffee',
    startDate: '2023-04-08',
    endDate: '2023-05-06',
  },
};

const mockHabitsData = [
  {
    id: 'a4e95781-9f45-4c38-8d67-21355fd5e228',
    habitDescription: 'Run 2 km every evening',
    habitType: 'develop',
    progressMetric: 'km',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
  {
    id: 'b5e95781-9f45-4c38-8d67-21355fd5e449',
    habitDescription: 'Drink only one coffee per day',
    habitType: 'break',
    progressMetric: 'count',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
];

const fourHabitsMockData = [
  {
    id: 'a4e95781-9f45-4c38-8d67-21355fd5e228',
    habitDescription: 'Run 2 km every evening',
    habitType: 'develop',
    progressMetric: 'km',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
  {
    id: 'b5e95781-9f45-4c38-8d67-21355fd5e449',
    habitDescription: 'Drink only one coffee per day',
    habitType: 'break',
    progressMetric: 'count',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
  {
    id: '66c8b782-6238-42a2-ab77-2c03d19b1d0f',
    habitDescription: 'Exercise daily',
    habitType: 'develop',
    progressMetric: 'minutes',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      startDate: '2023-04-08',
      endDate: '2023-05-06',
    },
  },
  {
    id: '87c8b782-6238-42a2-ab77-2c03d19b1d05n',
    habitDescription: 'Do 15 minutes of yoga in the evening',
    habitType: 'develop',
    progressMetric: 'minutes',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Improve sleep quality by exercising and drinking less coffee',
      startDate: '2023-04-08',
      endDate: '2023-05-06',
    },
  },
];

const editHabitData = {
  id: 'a4e95781-9f45-4c38-8d67-21355fd5e228',
  habitDescription: 'Run 20km every week',
  habitType: 'develop',
  progressMetric: 'km',
  goal: {
    id: 'bec49b7a-7819-453b-9895-347e3982acc5',
    goalDefinition:
      'Improve sleep quality by exercising and drinking less coffee',
    priority: 'secondary',
    startDate: '2023-05-07',
    endDate: '2023-07-15',
  },
};

export {
  editGoalData,
  editHabitData,
  fourHabitsMockData,
  storageJwtKey,
  mockJwt,
  mockGoalData1,
  mockGoalData2,
  mockGoalData3,
  mockGoalFormData,
  mockHabitData,
  mockHabitFormData,
  mockHabitsData,
  mockUserData,
};
