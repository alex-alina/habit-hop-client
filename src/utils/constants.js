export const baseUrl = 'http://localhost:3002';

export const localStorageJwtKey = 'userJwt';
