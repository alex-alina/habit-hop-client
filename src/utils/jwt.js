export const jwtData = (jwt) => {
  var base64Url = jwt.split('.')[1];
  var base64 = base64Url.replace('-', '+').replace('_', '/');
  return JSON.parse(window.atob(base64));
};

const currentTime = Date.now() / 1000;

export const extractUserId = (jwt) => jwtData(jwt).id;

export const isExpired = (jwt, now = currentTime) => jwtData(jwt).exp < now;
