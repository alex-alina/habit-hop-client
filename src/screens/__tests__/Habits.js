import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Router from 'react-router';
import { mockHabitFormData, mockJwt } from '../../mocks/constants';
import {
  deleteHabitException,
  deleteHabitNetworkError,
  deleteHabitServerDownError,
  getGoalException,
  getGoalNetworkError,
  getGoalServerDownError,
  getHabitsException,
  getHabitsNetworkError,
  getHabitsServerDownError,
  getUserException,
  getUserNetworkError,
  getUserServerDownError,
  getZeroHabitsHandler,
  getFourHabitsHandler,
} from '../../mocks/handlers';
import { server } from '../../mocks/server';
import * as jwtModule from '../../utils/jwt';
import { renderWithProvidersAndRouter } from '../../utils/testUtils';
import Habits from '../Habits';

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useParams: jest.fn(),
}));

//this mock goal starts with two habits (one develop and one break)
const goalId = 'bec49b7a-7819-453b-9895-347e3982acc5';

describe('Habits component', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => mockJwt),
        removeItem: jest.fn(() => null),
      },
      writable: true,
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders screen with goal listed in the side bar, and two habit cards', async () => {
    let goalsContainer;
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      const { container } = renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
      goalsContainer = container;
    });

    const habitCards = screen.getAllByText(/settings/i);
    const goalText = screen.getByText(
      /improve sleep quality by exercising and drinking less coffee/i
    );

    expect(goalsContainer).toBeInTheDocument();
    expect(goalsContainer.innerHTML).toMatchSnapshot();
    expect(habitCards).toHaveLength(2);
    expect(goalText).toBeInTheDocument();
  });

  it('allows a user with no habits to open the add habit form overlay', async () => {
    server.use(getZeroHabitsHandler);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
      reRender = rerender;
    });

    const user = userEvent.setup();
    const noHabitsIntro = screen.getByText(
      'Oops, you have no habits for this goal...'
    );

    expect(noHabitsIntro).toBeInTheDocument();
    expect(screen.queryByText('Habit description')).not.toBeInTheDocument();

    await user.click(screen.getByText(/add habit/i));
    expect(screen.getByLabelText('Habit description')).toBeInTheDocument();
    expect(screen.getByText('Add new habit')).toBeInTheDocument();

    const closeBtn = screen.getByLabelText('Close overlay');
    await userEvent.click(closeBtn);

    await act(async () => {
      reRender(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    expect(
      screen.queryByLabelText('Habit description')
    ).not.toBeInTheDocument();
  });

  it('allows the user to open and close the habit form overlay', async () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
      reRender = rerender;
    });

    const user = userEvent.setup();
    const formTextInput = screen.queryByLabelText('Habit description');
    expect(formTextInput).not.toBeInTheDocument();

    await user.click(screen.getByText(/add habit/i));
    expect(screen.getByLabelText('Habit description')).toBeInTheDocument();
    expect(screen.getByText('Add new habit')).toBeInTheDocument();

    const closeBtn = screen.getByLabelText('Close overlay');
    await userEvent.click(closeBtn);

    await act(async () => {
      reRender(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    expect(formTextInput).not.toBeInTheDocument();
  });

  it('allows the user to add a new habit if they have less than four habits', async () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
      reRender = rerender;
    });

    const user = userEvent.setup();
    const habitCards = screen.getAllByText(/settings/i);
    expect(habitCards).toHaveLength(2);

    await user.click(screen.getByText(/add habit/i));

    await user.type(
      screen.getByLabelText('Habit description'),
      mockHabitFormData.habitDescription
    );
    await user.click(
      screen.getByLabelText('develop new habit'),
      mockHabitFormData.habitType
    );
    await user.click(
      screen.getByLabelText('minutes'),
      mockHabitFormData.progressMetric
    );
    await user.click(screen.getByText('Add new habit'));

    await act(async () => {
      reRender(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });
    const habitOneText = screen.getByText(/run 2 km every evening/i);
    const habitTwoText = screen.getByText(/drink only one coffee per day/i);
    const habitThreeText = screen.getByText(mockHabitFormData.habitDescription);
    expect(screen.getAllByText(/settings/i)).toHaveLength(3);
    expect(habitOneText).toBeInTheDocument();
    expect(habitTwoText).toBeInTheDocument();
    expect(habitThreeText).toBeInTheDocument();
  });

  it('allows the user to delete a habit from the list', async () => {
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const user = userEvent.setup();
    const habitCards = screen.getAllByText(/settings/i);
    expect(habitCards).toHaveLength(2);
    expect(screen.queryByText(/delete/i)).not.toBeInTheDocument();

    const firstSettingsBtn = habitCards[0];
    await user.click(firstSettingsBtn);
    expect(screen.getByText(/delete/i)).toBeInTheDocument();

    await user.click(screen.getByText(/delete/i));
    expect(screen.getAllByText(/settings/i)).toHaveLength(1);
  });

  it('allows user to log out and redirects to /login', async () => {
    let spy;
    spy = jest.spyOn(jwtModule, 'isExpired');
    spy.mockReturnValue(false);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    expect(spy).toHaveBeenCalled();
    expect(window.location.pathname).toBe(`/habits/${goalId}`);
    const logOutBtns = screen.getAllByText(/log out/i);
    await userEvent.click(logOutBtns[0]);
    expect(window.location.pathname).toBe('/login');
  });

  it('logs out user and redirects to /login - second button', async () => {
    let spy;
    spy = jest.spyOn(jwtModule, 'isExpired');
    spy.mockReturnValue(false);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    expect(spy).toHaveBeenCalled();
    expect(window.location.pathname).toBe(`/habits/${goalId}`);
    const logOutBtns = screen.getAllByText(/log out/i);
    await userEvent.click(logOutBtns[1]);
    expect(window.location.pathname).toBe('/login');
  });

  it('logs out user and redirects to /login when token is expired', async () => {
    let spy;
    spy = jest.spyOn(jwtModule, 'isExpired');
    spy.mockReturnValue(true);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    expect(spy).toHaveBeenCalled();
    expect(window.location.pathname).not.toBe(`/habits/${goalId}`);
    expect(window.location.pathname).toBe('/login');
  });

  it('displays four habit cards and info banner', async () => {
    server.use(getFourHabitsHandler);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const banner = screen.getByText(
      /You can add a maximum of four habits per goal/i
    );
    expect(banner).toBeInTheDocument();
    expect(screen.getAllByText(/settings/i)).toHaveLength(4);
  });

  it('closes the info banner for user with three goals', async () => {
    server.use(getFourHabitsHandler);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const banner = screen.getByText(
      /You can add a maximum of four habits per goal/i
    );

    expect(banner).toBeInTheDocument();
    await userEvent.click(
      screen.getByRole('button', { name: /close banner/i })
    );

    expect(banner).not.toBeInTheDocument();
  });
});

describe('Habits component (errors)', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => mockJwt),
        removeItem: jest.fn(() => null),
      },
      writable: true,
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders screen with error messages: no goals, user or habits data', async () => {
    server.use(getUserException);
    server.use(getGoalException);
    server.use(getHabitsException);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });
    expect(screen.getByText('Test Error: User not found')).toBeInTheDocument();
    expect(screen.getByText('Test Error: Goal not found')).toBeInTheDocument();
    expect(
      screen.getByText('Test Error: Habits not found')
    ).toBeInTheDocument();
  });

  it('renders screen with server down error message', async () => {
    server.use(getUserServerDownError);
    server.use(getGoalServerDownError);
    server.use(getHabitsServerDownError);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const error = /There are issues with the server. Please try again later/i;
    expect(screen.getAllByText(error)[0]).toBeInTheDocument();
    expect(screen.getAllByText(error)).toHaveLength(3);
  });

  it('renders screen with network error message', async () => {
    server.use(getUserNetworkError);
    server.use(getGoalNetworkError);
    server.use(getHabitsNetworkError);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const error = /Test: some other network error/i;
    expect(screen.getAllByText(error)[0]).toBeInTheDocument();
    expect(screen.getAllByText(error)).toHaveLength(3);
  });

  it('renders server error when habit could not be deleted', async () => {
    server.use(deleteHabitException);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const user = userEvent.setup();
    const habitCards = screen.getAllByText(/settings/i);
    expect(habitCards).toHaveLength(2);

    const firstSettingsBtn = habitCards[0];
    await user.click(firstSettingsBtn);
    await user.click(screen.getByText(/delete/i));
    expect(screen.getAllByText(/settings/i)).toHaveLength(2);
    expect(
      screen.getByText('Test Error: Habit was not deleted')
    ).toBeInTheDocument();
  });

  it('renders server down error message when user tries to delete habit', async () => {
    server.use(deleteHabitServerDownError);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const user = userEvent.setup();
    const error = /There are issues with the server. Please try again later/i;
    const habitCards = screen.getAllByText(/settings/i);
    const firstSettingsBtn = habitCards[0];
    await user.click(firstSettingsBtn);
    await user.click(screen.getByText(/delete/i));
    expect(screen.getAllByText(/settings/i)).toHaveLength(2);
    expect(screen.getByText(error)).toBeInTheDocument();
  });

  it('renders network error message when user tries to delete habit', async () => {
    server.use(deleteHabitNetworkError);
    jest.spyOn(Router, 'useParams').mockReturnValue({ goalId });

    await act(async () => {
      renderWithProvidersAndRouter(<Habits />, {
        route: `/habits/${goalId}`,
      });
    });

    const user = userEvent.setup();
    const error = /Test: some other network error/i;
    const habitCards = screen.getAllByText(/settings/i);
    const firstSettingsBtn = habitCards[0];
    await user.click(firstSettingsBtn);
    await user.click(screen.getByText(/delete/i));
    expect(screen.getAllByText(/settings/i)).toHaveLength(2);
    expect(screen.getByText(error)).toBeInTheDocument();
  });
});
