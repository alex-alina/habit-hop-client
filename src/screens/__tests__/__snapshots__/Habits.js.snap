// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habits component renders screen with goal listed in the side bar, and two habit cards 1`] = `"<div display=\\"flex\\" width=\\"100%\\" class=\\"sc-gswNZR kUduhK\\"><div width=\\"auto,auto,auto,auto,25%\\" class=\\"sc-gswNZR dJMIkw\\"><div display=\\"flex\\" class=\\"sc-gswNZR vkNxI\\"><a href=\\"/goals\\"><button aria-label=\\"Back to goals page\\" display=\\"flex\\" height=\\"40\\" width=\\"40\\" class=\\"sc-bcXHqe gGHIzM\\"><svg width=\\"26px\\" height=\\"26px\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"arrow-left\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP fbRFTU\\"><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M5.79889 24H41.7989\\"></path><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M17.7988 36L5.79883 24L17.7988 12\\"></path></svg></button></a><button display=\\"block,block,block,block,none\\" class=\\"sc-bcXHqe jmzQEq\\">Log out</button></div><div display=\\"flex\\" class=\\"sc-gswNZR jrZubv\\"><div font-size=\\"4\\" width=\\"100%\\" class=\\"sc-gswNZR bDZAsr\\">Improve sleep quality by exercising and drinking less coffee</div><div color=\\"grey.4\\" font-size=\\"4\\" height=\\"60\\" width=\\"100%\\" class=\\"sc-gswNZR eCTDea\\">Filters container placeholder</div><div display=\\"none,none,none,none,block\\" class=\\"sc-gswNZR fzHayT\\"><svg>Data.svg</svg></div></div></div><div width=\\"auto,auto,auto,auto,75%\\" class=\\"sc-gswNZR dKljkg\\"><div display=\\"none,none,none,none,flex\\" width=\\"100%\\" class=\\"sc-gswNZR EEXWC\\"><button class=\\"sc-bcXHqe hXZpbj\\">Log out</button></div><div display=\\"flex\\" class=\\"sc-gswNZR eyzHEH\\"><h1 color=\\"heading\\" font-family=\\"heading\\" font-weight=\\"1\\" class=\\"sc-jSUZER fZmrhL\\">Habits' Overview</h1><div display=\\"flex\\" width=\\"90%,90%,90%,90%,70%\\" class=\\"sc-gswNZR hDPBwo\\"><div display=\\"flex\\" class=\\"sc-gswNZR kHkLCL\\"><button display=\\"flex\\" width=\\"150\\" class=\\"sc-bcXHqe EgaYZ\\">Settings<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"down-c\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M33 21L24 30L15 21\\"></path></svg></button><button display=\\"flex\\" width=\\"150\\" class=\\"sc-bcXHqe iWYaAY\\">New Entry<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"add-one\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 16V32\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M16 24L32 24\\"></path></svg></button></div><div display=\\"flex\\" class=\\"sc-gswNZR cJRnNB\\"><svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"check-one\\" aria-hidden=\\"true\\" role=\\"graphics-symbol\\" class=\\"sc-hLBbgP fRRZMG\\"><path fill=\\"none\\" stroke=\\"#1D8348\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C29.5228 44 34.5228 41.7614 38.1421 38.1421C41.7614 34.5228 44 29.5228 44 24C44 18.4772 41.7614 13.4772 38.1421 9.85786C34.5228 6.23858 29.5228 4 24 4C18.4772 4 13.4772 6.23858 9.85786 9.85786C6.23858 13.4772 4 18.4772 4 24C4 29.5228 6.23858 34.5228 9.85786 38.1421C13.4772 41.7614 18.4772 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#1D8348\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M16 24L22 30L34 18\\"></path></svg><p font-family=\\"body\\" font-size=\\"4\\" class=\\"sc-eDvSVe iNfMhQ\\">Run 2 km every evening</p></div><button display=\\"flex\\" width=\\"170\\" class=\\"sc-bcXHqe cqLsUE\\">Show entries<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"down-c\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M33 21L24 30L15 21\\"></path></svg></button></div><div display=\\"flex\\" width=\\"90%,90%,90%,90%,70%\\" class=\\"sc-gswNZR hDPBwo\\"><div display=\\"flex\\" class=\\"sc-gswNZR kHkLCL\\"><button display=\\"flex\\" width=\\"150\\" class=\\"sc-bcXHqe EgaYZ\\">Settings<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"down-c\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M33 21L24 30L15 21\\"></path></svg></button><button display=\\"flex\\" width=\\"150\\" class=\\"sc-bcXHqe iWYaAY\\">New Entry<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"add-one\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 16V32\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M16 24L32 24\\"></path></svg></button></div><div display=\\"flex\\" class=\\"sc-gswNZR cJRnNB\\"><svg width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"close-one\\" aria-hidden=\\"true\\" role=\\"graphics-symbol\\" class=\\"sc-hLBbgP fRRZMG\\"><path fill=\\"none\\" stroke=\\"#922B21\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#922B21\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M29.6567 18.3432L18.343 29.6569\\"></path><path fill=\\"none\\" stroke=\\"#922B21\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M18.3433 18.3432L29.657 29.6569\\"></path></svg><p font-family=\\"body\\" font-size=\\"4\\" class=\\"sc-eDvSVe iNfMhQ\\">Drink only one coffee per day</p></div><button display=\\"flex\\" width=\\"170\\" class=\\"sc-bcXHqe cqLsUE\\">Show entries<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"down-c\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#1A237E\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M33 21L24 30L15 21\\"></path></svg></button></div></div><button display=\\"flex\\" class=\\"sc-bcXHqe eILiBH\\">Add habit<svg width=\\"18\\" style=\\"margin-left: 8px;\\" height=\\"18\\" viewBox=\\"0 0 48 48\\" xmlns=\\"http://www.w3.org/2000/svg\\" name=\\"add-one\\" role=\\"graphics-symbol\\" aria-hidden=\\"true\\" class=\\"sc-hLBbgP ixtKLE\\"><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 44C35.0457 44 44 35.0457 44 24C44 12.9543 35.0457 4 24 4C12.9543 4 4 12.9543 4 24C4 35.0457 12.9543 44 24 44Z\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M24 16V32\\"></path><path fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"4\\" stroke-linejoin=\\"round\\" stroke-linecap=\\"round\\" d=\\"M16 24L32 24\\"></path></svg></button></div></div>"`;
